cf_xarray/geometry.py

```diff
@@ -1,11 +1,188 @@
 from __future__ import annotations
 
+import copy
 from collections.abc import Sequence
 
 import numpy as np
 import pandas as pd
 import xarray as xr
 
+GEOMETRY_CONTAINER_NAME = "geometry_container"
+
+__all__ = [
+    "decode_geometries",
+    "encode_geometries",
+    "cf_to_shapely",
+    "shapely_to_cf",
+]
+
+
+def decode_geometries(encoded: xr.Dataset) -> xr.Dataset:
+    """
+    Decode CF encoded geometries to a numpy object array containing shapely geometries.
+
+    Parameters
+    ----------
+    encoded : Dataset
+        A Xarray Dataset containing encoded geometries.
+
+    Returns
+    -------
+    Dataset
+        A Xarray Dataset containing decoded geometries.
+
+    See Also
+    --------
+    shapely_to_cf
+    cf_to_shapely
+    encode_geometries
+    """
+    if GEOMETRY_CONTAINER_NAME not in encoded._variables:
+        raise NotImplementedError(
+            f"Currently only a single geometry variable named {GEOMETRY_CONTAINER_NAME!r} is supported."
+            "A variable by this name is not present in the provided dataset."
+        )
+
+    enc_geom_var = encoded[GEOMETRY_CONTAINER_NAME]
+    geom_attrs = enc_geom_var.attrs
+    # Grab the coordinates attribute
+    geom_attrs.update(enc_geom_var.encoding)
+
+    geom_var = cf_to_shapely(encoded).variable
+
+    todrop = (GEOMETRY_CONTAINER_NAME,) + tuple(
+        s
+        for s in " ".join(
+            geom_attrs.get(attr, "")
+            for attr in [
+                "interior_ring",
+                "node_coordinates",
+                "node_count",
+                "part_node_count",
+                "coordinates",
+            ]
+        ).split(" ")
+        if s
+    )
+    decoded = encoded.drop_vars(todrop)
+
+    name = geom_attrs.get("variable_name", None)
+    if name in decoded.dims:
+        decoded = decoded.assign_coords(
+            xr.Coordinates(coords={name: geom_var}, indexes={})
+        )
+    else:
+        decoded[name] = geom_var
+
+    # Is this a good idea? We are deleting information.
+    for var in decoded._variables.values():
+        if var.attrs.get("geometry") == GEOMETRY_CONTAINER_NAME:
+            var.attrs.pop("geometry")
+    return decoded
+
+
+def encode_geometries(ds: xr.Dataset):
+    """
+    Encode any discovered geometry variables using the CF conventions.
+
+    Practically speaking, geometry variables are numpy object arrays where the first
+    element is a shapely geometry.
+
+    .. warning::
+
+       Only a single geometry variable is supported at present. Contributions to fix this
+       are welcome.
```

Comment:
I think that is actually a limitation of CF conventions, no? I have a vague memory that it supports at max one geometry variable.

Reply:
Each variable is only linked to one geometry variable through the `geometry` attribute. But I can't see why you can't have multiple geometry variables.